io/src/hash.rs

```diff
@@ -0,0 +1,387 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! I/O hashing support.
+//!
+//! Support for various hashing related things e.g.
+//!
+//! - Hashing to a writer.
+//! - Implement I/O traits for hash engines.
+
+use hashes::hmac::HmacEngine;
+use hashes::{
+    hash160, ripemd160, sha1, sha256, sha256d, sha256t, sha384, sha512, sha512_256, siphash24,
+    GeneralHash, HashEngine as _,
+};
+
+macro_rules! impl_write {
+    ($ty: ty, $write_fn: expr, $flush_fn: expr $(, $bounded_ty: ident : $bounds: path),*) => {
+        // `std::io::Write` is implemented in `bitcoin_hashes` because of the orphan rule.
+        impl<$($bounded_ty: $bounds),*> crate::Write for $ty {
+            #[inline]
+            fn write(&mut self, buf: &[u8]) -> crate::Result<usize> {
+                $write_fn(self, buf)}
```

Comment:
I would've liked overriding `write_all` as well.

Reply:
Promoted to #4034